### PR TITLE
pageserver: sharding followups

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

This PR contains pageserver changes made while shaking out and testing sharding, including shard splitting.  The API for shard splitting will come later, the related changes in this PR are enablers for that.

**Recommended to review this commit-by-commit since it's a series of small independent changes**

## Summary of changes

Enabling changes for shard splitting:
- Don't physically delete layers when removing from index, if they belong to an ancestor shard
- During compaction, don't retain keys if they are present in a layer but no longer belong to the current ShardIdentity.  This avoids generating image layers with many defunct keys that inflate the size of a post-split shard.

Fixes for running multiple shards on the same node (this is permitted in general but comes up especially when splitting):
- Fix a case where shard lookup could hit a secondary for a tenant and incorrectly conclude no shard found
- Fix a case where page service would wait for the desired LSN on a different timeline than it subsequently read, resulting in incorrect read results.

Quality of life changes:
- If a request hits the wrong shard, log more detail about it.
- Make walredo code shard aware, so that its logging includes shard IDs (functionally no change)
- Tweak background runtime initialization to always create at least two threads: otherwise running a pageserver pinned to a single core fails.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by jcsp** [comment]
`test_large_schema` failure was an assertion in walproposer.c, so highly unlikely to be related to this PR (https://github.com/neondatabase/neon/issues/6248)

File: pageserver/src/task_mgr.rs
Comment (knizhnik) [inline pageserver/src/task_mgr.rs:150]: Does it have any relation to sharding and this PR?

File: pageserver/src/tenant.rs
Comment (knizhnik) [inline pageserver/src/tenant.rs:598]: Why do we need to make weal redo shard-aware?
It seems to me that walredo should not just be called for pages not belonging to the shard..

File: pageserver/benches/bench_walredo.rs
Comment (knizhnik) [inline pageserver/benches/bench_walredo.rs:30]: It will be nice to explain what does `TenantShardId::unsharded` mean.
I looks in shard.rs at the sources of this method and still the meaning of this name is obscure to me...

File: libs/pageserver_api/src/keyspace.rs
Comment (knizhnik) [inline libs/pageserver_api/src/keyspace.rs:127]: May be I missed something, but it seems to me that for huge stripes excluding from key space ranges not belonging to the shard may be even more critical.

**Review by knizhnik — APPROVED** [review]

File: pageserver/src/tenant.rs
Comment (jcsp) [reply pageserver/src/tenant.rs:598]: See description:
> Make walredo code shard aware, so that its logging includes shard IDs (functionally no change)

File: pageserver/src/task_mgr.rs
Comment (jcsp) [reply pageserver/src/task_mgr.rs:150]: Not directly - it came up in the context of testing sharding because I was working with a neon_local setup where I had many pageservers pinned to cores.

**Review by problame — COMMENTED** [review]
The warning log of `Page request routed to wrong shard:` could easily flood the log should it ever happen in prod. Did you consider this? (The anyhow::anyhow! also isn't great because of CPU overhead, but, I guess that'll be solved in #6038)

Same concern for the warning added to   `Timeline::get` `get on disposable key `

---

Regarding 1ce33ad9d720733bd4c96791c8321ad10ebb99e5 : can't spot what exactly was broken. Commit message should elaborate.

---

Regarding the `get on disposable key`: why can't we bail out there with a `PageReconstructError`?

--- 

Regarding ba4cbc651070243a9ec911e0cb497c22e514faeb , I don't understand how this can happen (probably because I don't understand how splits happen). 
I would have expected that splits never happen in place, but, instead after a split, each shard has its own tenant-$shard_id directory with its own layer files.

---

Please make sure to hit "merge by rebase" instead of "merge by squash" when this gets merged.

**Comment by jcsp** [comment]
> Regarding 1ce33ad : can't spot what exactly was broken. Commit message should elaborate.

Updated it as follows:

pageserver: fix handling getpage with multiple shards on one node

Previously, we would wait for the LSN to be visible on whichever
timeline we happened to load at the start of the connection, then
proceed to look up the correct timeline for the key and do the read.

If the timeline holding the key was behind the timeline we used
for the LSN wait, then we might serve an apparently-successful read result
that actually contains data from behind the requested lsn.

**Comment by jcsp** [comment]
> The warning log of Page request routed to wrong shard: could easily flood the log should it ever happen in prod. Did you consider this? (The anyhow::anyhow! also isn't great because of CPU overhead, but, I guess that'll be solved in https://github.com/neondatabase/neon/issues/6038)

For this one, in #6038 I think we'll make this disconnect the client, so that the postgres-side backoff on disconnect will implicitly rate limit it.

> Same concern for the warning added to Timeline::get get on disposable key 
> Regarding the get on disposable key: why can't we bail out there with a PageReconstructError?

I took another look at this path.  The warning I added there was helpful when debugging an issue looking up timelines, but I don't think we need to spend the cycles on it in prod, because it is redundant wrt. the is_key_local check we do when looking up the Timeline from page_service.  I've downgraded this to a debug_assert!().

**Comment by jcsp** [comment]
> I would have expected that splits never happen in place, but, instead after a split, each shard has its own tenant-$shard_id directory with its own layer files.

That is the case locally (shards never touch each others files).  But in remote storage, the child shards contain references to their parents' layers -- the purpose  of IndexPart containing the shard index in each layer's metadata is to enable that.

File: pageserver/benches/bench_walredo.rs
Comment (jcsp) [reply pageserver/benches/bench_walredo.rs:30]: I've tacked that onto the list of things to clean up in the names/docs of helpers https://github.com/neondatabase/neon/issues/6072

**Comment by problame** [comment]
> > I would have expected that splits never happen in place, but, instead after a split, each shard has its own tenant-$shard_id directory with its own layer files.
> 
> That is the case locally (shards never touch each others files). But in remote storage, the child shards contain references to their parents' layers -- the purpose of IndexPart containing the shard index in each layer's metadata is to enable that.

Ok, gotcha. It's obvious to me now, wonder why it wasn't in review.
Maybe I didn't have the full context of `struct Layer` loaded into my head.
Or it was something in the comment, but, can't pinpoint it.

It would be nice to have a regression test for this behavior though, it's quite essential.

**Comment by jcsp** [comment]
> It would be nice to have a regression test for this behavior though, it's quite essential.

Definitely: we just can't land any sharding tests until the compute support lands (currently all the testing is happening on a branch with the compute commits in it.)

**Pull request merged**
